server/handlers/validation.go

```diff
@@ -125,34 +77,150 @@ func validateUpdate(gun string, updates []storage.MetaUpdate, store storage.Meta
+		logrus.Debug("Successfully validated snapshot")
+	} else {
+		// Check:
+		//	 - we have a snapshot key
+		//   - it matches a snapshot key signed into the root.json
+		// Then:
+		//   - generate a new snapshot
+		//   - add it to the updates
+		//
+		// TODO: Assumptions have been made about what is loaded into the repo.
+		//       we need to ensure at a minimum the root and targets are loaded
+		//       in.
+		err := prepRepo(gun, repo, store)
+		if err != nil {
+			return nil, err
 		}
+		update, err := generateSnapshot(gun, kdb, repo, store)
+		if err != nil {
+			return nil, err
+		}
+		updates = append(updates, *update)
 	}
+	return updates, nil
+}
 
-	if err := validateSnapshot(snapshotRole, oldSnap, roles[snapshotRole], roles, kdb); err != nil {
-		logrus.Error("ErrBadSnapshot: ", err.Error())
-		return ErrBadSnapshot{msg: err.Error()}
+func prepRepo(gun string, repo *tuf.Repo, store storage.MetaStore) error {
+	if repo.Root == nil {
+		rootJSON, err := store.GetCurrent(gun, data.CanonicalRootRole)
+		if err != nil {
+			return fmt.Errorf("could not load repo for snapshot generation: %v", err)
+		}
+		root := &data.SignedRoot{}
+		err = json.Unmarshal(rootJSON, root)
+		if err != nil {
+			return fmt.Errorf("could not load repo for snapshot generation: %v", err)
+		}
+		repo.SetRoot(root)
+	}
+	if repo.Targets[data.CanonicalTargetsRole] == nil {
+		targetsJSON, err := store.GetCurrent(gun, data.CanonicalTargetsRole)
+		if err != nil {
+			return fmt.Errorf("could not load repo for snapshot generation: %v", err)
+		}
+		targets := &data.SignedTargets{}
+		err = json.Unmarshal(targetsJSON, targets)
+		if err != nil {
+			return fmt.Errorf("could not load repo for snapshot generation: %v", err)
+		}
+		repo.SetTargets(data.CanonicalTargetsRole, targets)
 	}
-	logrus.Debug("Successfully validated snapshot")
 	return nil
 }
 
+func generateSnapshot(gun string, kdb *keys.KeyDB, repo *tuf.Repo, store storage.MetaStore) (*storage.MetaUpdate, error) {
+	role := kdb.GetRole(data.RoleName(data.CanonicalSnapshotRole))
+	if role == nil {
+		return nil, ErrBadRoot{msg: "root did not include snapshot role"}
+	}
+
+	algo, keyBytes, err := store.GetKey(gun, data.CanonicalSnapshotRole)
+	foundK := data.NewPublicKey(algo, keyBytes)
+
+	validKey := false
+	for _, id := range role.KeyIDs {
+		if id == foundK.ID() {
+			validKey = true
+			break
+		}
+	}
+	if !validKey {
+		return nil, ErrBadHierarchy{msg: "no snapshot was included in update and server does not hold current snapshot key for repository"}
+	}
+
+	currentJSON, err := store.GetCurrent(gun, data.CanonicalSnapshotRole)
+	if err != nil {
+		if _, ok := err.(*storage.ErrNotFound); !ok {
+			return nil, fmt.Errorf("could not retrieve previous snapshot: %v", err)
+		}
+	}
+	var sn *data.SignedSnapshot
+	if currentJSON != nil {
+		sn = new(data.SignedSnapshot)
+		err := json.Unmarshal(currentJSON, sn)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve previous snapshot: %v", err)
```

Comment:
It's indeterminate whether it's recoverable. Clients will reject the new snapshot if its version is not greater than the previous one. If we can't parse the previous one, we don't know what the version of the new one should be.

We have some yet to be done work around resetting the version counter for a role when its key gets rotated which would potentially make this a recoverable state, but for now, I don't think there's much we can do short of manual resolution.